tower-balance/src/p2c/service.rs

```diff
@@ -307,25 +328,62 @@ where
 }
 
 impl<K, S: Service<Req>, Req> Future for UnreadyService<K, S, Req> {
-    type Item = (K, S);
-    type Error = (K, Error<S::Error>);
+    type Output = Result<(K, S), (K, Error<S::Error>)>;
 
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        if let Ok(Async::Ready(())) = self.cancel.poll() {
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let me = self.project();
+
+        if let Poll::Ready(Ok(())) = me.cancel.poll(cx) {
             let key = self.key.take().expect("polled after ready");
-            return Err((key, Error::Canceled));
+            return Poll::Ready(Err((key, Error::Canceled)));
         }
 
-        match self.ready.poll() {
-            Ok(Async::NotReady) => Ok(Async::NotReady),
-            Ok(Async::Ready(svc)) => {
+        match me.ready.poll(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Ok(svc)) => {
                 let key = self.key.take().expect("polled after ready");
-                Ok((key, svc).into())
+                Ok((key, svc)).into()
             }
-            Err(e) => {
+            Poll::Ready(Err(e)) => {
                 let key = self.key.take().expect("polled after ready");
-                Err((key, Error::Inner(e)))
+                Err((key, Error::Inner(e))).into()
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Ready<S, Req> {
+    inner: Option<S>,
+    _pd: PhantomData<Req>,
+}
+
+impl<S, Req> Ready<S, Req> {
+    pub fn new(inner: S) -> Self {
+        Self {
+            inner: Some(inner),
+            _pd: PhantomData,
+        }
+    }
+}
+
+impl<S, Req> Unpin for Ready<S, Req> {}
```

Comment:
Probably not okay.

Reply:
So since poll_ready just takes a `&mut self` it must be Unpin therefore any !Unpin future within it must be Boxed etc. This makes this Unpin safe. So from what I understand if you have _no_ unsafe projections then impl Unpin is 100% safe.

Reply:
Right, but then why do you need this `impl` at all?

Reply:
So I need the impl to deref Ready -> &mut Ready so that I can poll the inner type. Since S is always Unpin there are two ways to do this:
- safely impl Unpin
- S: Unpin so that the compiler can auto generate the same as above. 

so by adding impl Unpin in this case I am avoiding adding + Unpin everywhere pretty much.

Reply:
I don't follow -- why is `S` always `Unpin`?

Reply:
Technically speaking it is not, but since we only need a &mut T and it is never pinned in the first place, it acts as if it were Unpin always since we already have a &mut T. The thing with Unpin is that it says that Pin<&mut T> is == to &mut T iff T: Unpin. So thus any &mut T is technically already Unpinned in theory.